<a href="https://colab.research.google.com/github/GoingForward1/practic_6_Pervuhin/blob/master/Lab_2_Pervuhin_(ML).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Блок 1. NumPy и векторизованная обработка данных.**

###**Задание 1.1.** Базовые свойства массивов

In [37]:
import numpy as np

delivery_time = np.array([35, 42, 28, 51, 39, 63, 31, 45])
order_price = np.array([1200, 3500, 800, 4200, 1900, 5600,
1500, 2800])
items_count = np.array([2, 5, 1, 7, 3, 8, 2, 4])

Использование shape, ndim, size:

In [38]:
print("delivery_time:", delivery_time.shape, delivery_time.ndim, delivery_time.size)
print("order_price:",order_price.shape, order_price.ndim, order_price.size)
print("items_count:",items_count.shape, items_count.ndim, items_count.size)

delivery_time: (8,) 1 8
order_price: (8,) 1 8
items_count: (8,) 1 8


*shape - выводит кортеж с количеством рядов в каждом измерении, ndim - кол-во измерений массива, size - кол-во элементов в массиве.*

*Каждый индекс должен соотвествовать одному и тому же заказу, чтобы было легко сопоставить данные об этом заказе зная всего лишь его индекс.*


*Вопрос для анализа: Если delivery_time содержит 8 наблюдений, а order_price всего 7, то мы теряем информацию о цене одного из заказов и получаем пропуск, что делает данные некачественными.*





###**Задание 1.2.** Двумерные массивы


In [39]:
import numpy as np

orders = np.array([
    [35, 1200, 2],
    [42, 3500, 5],
    [28, 800, 1],
    [51, 4200, 7],
    [39, 1900, 3],
    [63, 5600, 8],
    [31, 1500, 2],
    [45, 2800, 4]
])

1.Выведите orders.shape.

In [40]:
print(orders.shape)

(8, 3)


2.С помощью индексации получите:
*   данные первого заказа;
*   время доставки всех заказов;
*   стоимость всех заказов;
*   количество товаров в шестом заказе.

In [5]:
print("Данные первого заказа:", orders[0, :])
print("Время доставки всех заказов:",orders[:, 0])
print("Стоимость всех заказов:", orders[:, 1])
print("Кол-во товаров в 6 заказе:", orders[5, 2])

Данные первого заказа: [  35 1200    2]
Время доставки всех заказов: [35 42 28 51 39 63 31 45]
Стоимость всех заказов: [1200 3500  800 4200 1900 5600 1500 2800]
Кол-во товаров в 6 заказе: 8


3.С помощью срезов получите:
*   все заказы начиная с третьего;
*   первые пять заказов и первые два признака.



In [6]:
print("Все заказы начиная с 3-го:\n", orders[2:, :])
print("\nПервые 5 заказов и первые 2 признака:\n", orders[:5, :2])

Все заказы начиная с 3-го:
 [[  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]

Первые 5 заказов и первые 2 признака:
 [[  35 1200]
 [  42 3500]
 [  28  800]
 [  51 4200]
 [  39 1900]]


*Вывод: Можно вывести необходимую информацию в двумерных массивах при помощи обращении по индексам.*

*Вопрос для анализа: Объясните, почему: orders.shape возвращает (8, 3), а orders.size возвращает 24?*

*Ответ: Потому что метод shape возвращает размерность, а метод size кол-во элементов в массиве*

###**Задание 1.3.** View и Copy

In [7]:
delivery = orders[:, 0]

1.Измените первый элемент delivery на 1000.

In [8]:
delivery[0] = 1000
print(delivery[0])

1000


2.Выведите orders.

In [9]:
print(orders)

[[1000 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]


3.Объясните, почему изменение delivery повлияло на исходную матрицу.

*Изменение delivery повлияло на исходную матрицу, потому что они оба ссылаются на один и тот же участок в памяти. Сделано это для экономии оперативной памяти.*


4.Восстановите исходную матрицу orders.

In [10]:
orders[0,0] = 35
print(orders)

[[  35 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]


5-7.Создайте независимую копию delivery_copy. Измените первый элемент delivery_copy. Снова выведите orders

In [11]:
delivery_copy = orders[:, 0].copy()

delivery_copy[0] = 1000

print(orders)

[[  35 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]


8.Сравните результаты двух экспериментов.

*Без использования метода copy исходный массив order менялся при соответсвующем изменении массива delivery. C методом copy создалась копия стобца order и при его изменении сам массив order не менялся, так как ссылался на другой участок памяти.*

Вопросы для анализа:

Почему понимание различия между view и copy важно при обработке реальных наборов
данных?

*Потому что можно случайно изменить или удалить исходные данные, что может сделать набор данных нерелевантным.*

Приведите пример ситуации, в которой случайное изменение view может привести к ошибке в дальнейшем анализе.

*Добавить к одному или нескольким записям в одном признаке сильно превышающее значение может привести к перекосу данных к недействительным значениям и как результат будущая модель будет склонна делать неверные результаты.*


###**Задание 1.4.** Фильтрация с помощью булевых масок

In [12]:
delivery_time = np.array([35, 42, 28, 51, 39, 63, 31, 45])

1-2.Создайте булеву маску для поиска заказов с временем доставки больше 45 минут. Выведите маску.

In [13]:
mask_time = delivery_time>45
print(mask_time)

[False False False  True False  True False False]


3.Используйте маску для получения времени доставки проблемных заказов

In [14]:
filter_delivery = delivery_time[mask_time]
print(filter_delivery)

[51 63]


4.Определите количество таких заказов

In [15]:
count_filter = filter_delivery.size
print(count_filter)

2


5.Рассчитайте среднее время доставки для этой группы.

In [16]:
mean_filter = filter_delivery.mean()
print(mean_filter)

57.0


Вопросы для анализа:
В чём различие между условиями: delivery_time > 45 и delivery_time >= 45?

*В строгости: в первом случае маска для строго больше 45 минут, а во втором начиная от 45 минут.*

Как изменение граничного условия может повлиять на бизнес-метрику?

*Включить приемлемые данные как неблагоприятные (или наоборот) могут испортить качества оценки предоставления услуг и как следствие навредить развитию бизнеса*

###**Задание 1.5.** Фильтрация по нескольким условиям.

Менеджер продукта хочет изучить заказы, которые одновременно являются крупными по стоимости и содержат большое количество товаров.

1.Создайте единую булеву маску, учитывающую условия:
*   стоимость заказа больше 3000 рублей;
*   количество товаров больше 4



In [17]:
mask = (orders[:, 1] >3000) & (orders[:, 2] > 4)
print(mask)

[False  True False  True False  True False False]


2-3.Получите стоимость подходящих заказов.

In [18]:
expensive_orders = orders[mask]

print((expensive_orders)[:, 1])
print(expensive_orders[:, 0])

[3500 4200 5600]
[42 51 63]


4.Рассчитайте среднее время доставки выбранной группы.


In [19]:
print(expensive_orders[:, 0].mean())

52.0


Вопросы для анализа:
Почему анализ комбинации нескольких признаков может дать более полезный результат для бизнеса, чем анализ каждого признака отдельно?

*Потому что можно увидеть коррелляцию между различными признаками и сделать выводы от чего зависит выгода и качество услуг.*


##**Блок 2. Описательная статистика**

###**Задание 2.1.** Влияние выбросов на статистические показатели

Вы анализируете время доставки пяти заказов:

In [20]:
delivery_time = np.array([8, 9, 10, 11, 12])

1. Рассчитайте:
*   среднее арифметическое;
*   медиану

In [21]:
print("Среднее:", np.mean(delivery_time), "Медиана:", np.median(delivery_time))

Среднее: 10.0 Медиана: 10.0


2.Добавьте в выборку один аномальный результат — доставку продолжительностью
1000 минут.

In [22]:
delivery_time_corrupted = np.array([8, 9, 10, 11, 12, 1000])

3.Повторно рассчитайте среднее и медиану.

In [23]:
print("Среднее:", np.mean(delivery_time_corrupted), "Медиана:", np.median(delivery_time_corrupted))

Среднее: 175.0 Медиана: 10.5


Вопросы для анализа:

Почему среднее так сильно изменилось?

*Потому что его сместил выброс*

Почему медиана изменилась значительно меньше?

*Потому что она берет значение, которое находится в середине отсортированного списка всех этих чисел или среднее арифметическое между двумя по середине. В нашем случае между 10 и 11.*

Какой показатель лучше описывает типичное время доставки при наличии редких
экстремальных значений?

*Медиана благодаря тому что почти не изменяется при наличии выбросов.*

###**Задание 2.2.** Процентили и SLA

Интернет-магазин заявляет:
Не менее 90% заказов доставляются за 40 минут или быстрее.
Для проверки SLA смоделируйте 1000 наблюдений времени доставки.
Используйте нормальное распределение со следующими параметрами:

• среднее — 30 минут;

• стандартное отклонение — 5 минут.


In [24]:
np.random.seed(42)
delivery_simulated = np.random.normal(
    loc=30,
    scale=5,
    size=1000
)
print(delivery_simulated[:10]) #Первые 10 чисел

[32.48357077 29.30867849 33.23844269 37.61514928 28.82923313 28.82931522
 37.89606408 33.83717365 27.65262807 32.71280022]


 Рассчитайте:


*   среднее значение;
*   стандартное отклонение;
*   медиану.

In [25]:
print("Среднее:", np.mean(delivery_simulated), "Отклонение:", np.std(delivery_simulated), "Медиана:", np.median(delivery_simulated))


Среднее: 30.096660279111628 Отклонение: 4.893631038736771 Медиана: 30.12650306117444


Рассчитайте следующие процентили:


*   p50
*   p90
*   p95
*   p99


In [26]:
p = [0.5, 0.90, 0.95, 0.99]
np.quantile(delivery_simulated, p)

array([30.12650306, 36.52822601, 38.38486264, 41.5798475 ])

3.Сравните полученные значения с параметрами, использованными при генерации
данных.

*50% данных ниже ~30 и 50% данных выше этого значения - это соответсвует указанному loc в методе normal для delivery_simulated.*

4.Сравните полученные значения с параметрами, использованными при генерации
данных.
Используя p90, определите, выполняется ли заявленный SLA.


*90% всех заказов доставляются меньше чем за ~36.5 минут, что и входит в 40 минут заявленные в SLA.*

###**Задание 2.3.** Разброс и предсказуемость

Две курьерские службы выполнили по пять одинаковых доставок

In [27]:
service_A = np.array([28, 29, 30, 31, 32])
service_B = np.array([15, 20, 30, 40, 45])

1.Для каждой службы рассчитайте:
*   среднее;
*   медиану;
*   стандартное отклонение;
*   90-й процентиль.

In [28]:
print(f"Сервис А:\nСреднее:{np.mean(service_A)}  Медиана: {np.median(service_A)}  Стандартное отклонение: {np.std(service_A)}  90-й процентиль:  {np.quantile(service_A, 0.9)}")
print(f"Сервис B:\nСреднее:{np.mean(service_B)}  Медиана: {np.median(service_B)}  Стандартное отклонение: {np.std(service_B)}  90-й процентиль:  {np.quantile(service_B, 0.9)}")

Сервис А:
Среднее:30.0  Медиана: 30.0  Стандартное отклонение: 1.4142135623730951  90-й процентиль:  31.6
Сервис B:
Среднее:30.0  Медиана: 30.0  Стандартное отклонение: 11.40175425099138  90-й процентиль:  43.0


Вопросы для анализа:
Какая служба является более предсказуемой?

*Служба A - у нее меньше разброс значений от среднего.*

###**Задание 2.4.** Корреляция


Сгенерируйте данные:


In [29]:
np.random.seed(42)
items = np.random.randint(1, 10, 100)
delivery_time_correlated = (
    20
    + 3 * items
    + np.random.normal(0, 3, 100)
)

1.Рассчитайте коэффициент корреляции Пирсона между items и
delivery_time_correlated с помощью np.corrcoef()

In [30]:
matr_corr = np.corrcoef(items, delivery_time_correlated)
print(matr_corr)

[[1.         0.93432937]
 [0.93432937 1.        ]]


2.Извлеките коэффициент r

In [31]:
r = matr_corr[1,0]
print(r)

0.9343293664033214


3.Определите:
*   направление связи;
*   силу линейной связи

In [32]:
p = 1

if r > 0.7:
  print("Линейная связь сильная (>0.7) и прямая")
elif r > 0.3:
  print("Линейная связь средняя (>0.3) и прямая")
elif r > 0:
  print("Линейная связь слабая и прямая")
elif r == 0:
  print("Линейная связь отсутсвует")
elif r > -0.3:
  print("Линейная связь слабая и обратная")
elif r > -0.7:
  print("Линейная связь средняя и обратная")
else:
  print("Линейная связь сильная и обратная")


Линейная связь сильная (>0.7) и прямая


4. Сформулируйте возможное объяснение наблюдаемой зависимости с точки зрения
работы склада и доставки

*Чем больше товаров в заказе, тем больше нужно приготовить и подготовить заказ к доставке. К тому же, если доставка пешая или на обычном велосипеде, то от курьера требуется больше физических усилий, чтобы быстро доставить заказ из-за увеличения веса при увеличении кол-во товаров в заказе*

Вопрос для анализа:

Можно ли на основании высокого значения корреляции утверждать:
«Увеличение количества товаров является причиной увеличения времени доставки»

*Да. Пояснение было дано в блоке выше.*

###**Задание 2.5.** Корреляция и скрытые факторы

Рассмотрим ситуацию:
Летом одновременно увеличиваются:
• продажи мороженого;
• количество несчастных случаев на воде.
На основании этих данных аналитик делает вывод:
«Продажа мороженого приводит к увеличению количества несчастных случаев на воде»

1-3.Объясните, почему высокая корреляция не доказывает наличие причинно-следственной
связи. Определите фактор, который может влиять одновременно на оба показателя. Объясните механизм возникновения наблюдаемой корреляции

*Потому что как в данном примере может существовать третий скрытый признак/фактор с которым на самом деле и связаны оба явления - более жаркая погода летом чем в любой другой сезон. Тем самым они связанны через фактор-посредник, но не имеют причинно-следственной связи.*

##**Блок 3. Bootstrap и статистические гипотезы**

##**Задание 3.1.** Bootstrap и доверительный интервал

На складе измерили время сборки шести заказов

In [33]:
x = np.array([28, 31, 29, 45, 32, 30])

1.Рассчитайте среднее значение исходной выборки

In [34]:
mean_x = np.mean(x)
print(mean_x)

32.5


2.Реализуйте Bootstrap с количеством итераций.

B = 5000

На каждой итерации:
*   сформируйте выборку такого же размера, как x;
*   используйте выборку с возвращением;
*   рассчитайте её среднее;
*   сохраните результат
.

In [35]:
B = 5000
x_mean_lst = np.zeros(B)
for i in range(B):
  iter = np.random.choice(x, size = len(x), replace = True)
  x_mean_lst[i] = np.mean(iter)
print(x_mean_lst)

[30.5        32.         34.83333333 ... 30.16666667 32.5
 33.5       ]


Получите центральный 95%-й доверительный интервал для среднего

In [36]:
down = np.quantile(x_mean_lst, 0.025)
up = np.quantile(x_mean_lst, 0.975)
print("Доверительный интервал:", down, "-", up)

Доверительный интервал: 29.166666666666668 - 37.666666666666664


Вопросы для анализа:

Почему Bootstrap используется именно с выборкой с возвращением?

*Потому что мы не имеем настоящей генеральной совокупности, поэтому мы имитируем ее делая случайную выборку с возвращением. Если бы мы делали без возвращения - мы получали бы каждый раз исходную выборку*

 Как наличие наблюдения 45 влияет на Bootstrap-распределение среднего?

 *45 - выборос. В каких то Bootstrap-выборках он может встретиться ни разу а в каких то множество раз из-за чего стандартное отклонение Bootsrap-распределения среднего будет больше, а значит и сильнее разброс.*

  Почему результаты Bootstrap могут заметно изменяться при использовании очень
маленькой исходной выборки?

*Из-за разницы в величине данных*
Можно ли утверждать:
«Вероятность того, что истинное среднее находится внутри рассчитанного 95%-го
доверительного интервала, равна 95%»

*Нет. Истинное среднее - это постоянная. Если множества раз повторно сформировать выборку, то истинное среднее будет находится в этом интервалев 95% случаях.*

###**Задание 3.2.** Формулировка гипотез и ошибки I и II рода

Продуктовая команда внедрила новый алгоритм распределения заказов между курьерами.
Цель алгоритма — уменьшить среднее время доставки.
Обозначим:
*   \mu_A — истинное среднее время доставки в контрольной группе;
*   \mu_B — истинное среднее время доставки в группе с новым алгоритмом.

1-2.Сформулируйте нулевую гипотезу H_0. Сформулируйте альтернативную гипотезу H_1.

*H_0: \mu_A = \mu_B.*

*H_1: \mu_A < \mu_B*

Объясните применительно к этому эксперименту:
*   что означает ошибка I рода;
*   что означает ошибка II рода.

*Ошибка I рода: H_0 - верна, но мы ее отклоняем и считаем что алгоритм работает.
Ошибка I рода: H_0 - неверна, но мы ее не отклоняем и считаем что алгоритм не уменьшает время доставки.*

Укажите, какая ошибка соответствует:
* ложной тревоге;
* пропущенному эффекту

*Ложная тревога - ошибка I рода*

*Пропущеный эффект - ошибка II рода*

Полноценная эксплуатация нового алгоритма обходится компании в 5 млн рублей в год.

Какая ошибка может привести к прямым напрасным расходам компании — ошибка I или
ошибка II рода?

Обоснуйте ответ.

*Ошибка I рода может привести к напрасных расходам из-за бесполезного алгоритма, ведь он не работает и деньги на его внедрение выделены*

###**Задание 3.3.** p-value и принятие решения

После проведения A/B-теста статистический критерий дал:

p-value = 0.001

Уровень значимости:

α = 0.05

1.Дайте определение p-value

*p-value - вероятность получить наблюдаемое значение статистического критерия при условии что нулевая гипотеза верна.*

2.Объясните, почему утверждение:

«p-value = 0.001 означает, что вероятность истинности нулевой гипотезы равна 0.1%»

является неверным

*Нулевая гипотеза либо верна, либо нет (т.е. ее вероятность 1 или 0). P-value - это вероятность наблюдаемых данных ПРИ УСЛОВИИ что гипотеза верна, а не ее вероятность*

3.Определите, является ли результат статистически значимым при $\alpha = 0.05$

*Да, является. Так как 0.001 < 0.05, нулевая гипотеза отклоняется.*

Ситуация для анализа:

Предположим, эксперимент проведён на 1 000 000 пользователей.
Новый алгоритм уменьшил среднее время доставки всего на 3 секунды.
При этом статистический критерий дал:
p-value = 0.001
Эксплуатация нового алгоритма стоит компании 5 млн рублей в год.

Вопросы:

Является ли результат статистически значимым?

*Да. Т.к. p-value < 0.5*

Достаточно ли одного p-value для принятия решения о внедрении?

*Нет. p-value говорит лишь то, что эффект от алгоритма есть, но не какая у него сила и финансовая выгода*

Почему статистическая значимость не обязательно означает практическую или бизнес
значимость?

*При больших объемах выборки статистические тесы выявляют даже ничтожно малые различия. Как в данной ситуации: на выборке в 1000000 пользователей определили, что среднее время доставки уменьшилась на 3 секунды. Такие изменения слишком малы (особенно за 5млн рублей) чтобы принести выгоды бизнесу, поэтому у алгоритма очень низкая практическая значимость.*

Какие дополнительные показатели необходимо учитывать перед раскаткой алгоритма
на всех пользователей?

*Силу эффекта и соотношение прибыли от алгоритма/трат на внедрение*


Итоговый продуктовый вывод
Представьте, что вы отвечаете за принятие решения о внедрении алгоритма.
Сформулируйте итоговую рекомендацию и обоснуйте её.

*Выделять деньги на алгоритм и вводить его не целесообразно.*

*Статистическая значимость: Алгоритм действительно уменьшает время доставки (p-value = 0.0001).*

*Размер и практическая значимость эффекта: Размер эффекта равен всего 3 секундам, что абсолютно не заметно для бизнеса и пользователей.*

*Экономическая целесообразность: Сокращение доставки на 3 секунды при помощи алгоритма не приносит ощутимой финансовой выгоды, однако стоит его внедрение 5 млн рублей. Это просто напрасные расходы бизнеса.*

Представьте, что вы получили набор данных о работе интернет-магазина.
Среднее время доставки уменьшилось после изменения алгоритма, но одновременно:

• вырос разброс времени доставки;

• p90 практически не изменился;

• p95 увеличился;

• статистический тест показал p-value < 0.05.

Сформулируйте, какие дополнительные вопросы вы зададите перед тем, как рекомендовать
внедрение изменения на всех пользователей.

В ответе отдельно рассмотрите:

• качество данных;

• размер эффекта;

• распределение результатов;

• выбросы;

• процентили;

• статистическую значимость;

• практическую значимость;

• бизнес-эффект.

*Качество данных: Были ли исключены тех. сбои при получении времени доставки?*

*Размер эффекта: насколько снизилось среднее время доставки?*

*Распределение результатов: Появились ли тяжелые хвосты?*

*Выбросы: Почему появился такой резкий рост разброса значений?*

*Процентили: Почему p95 увиличелся, а p90 нет?*

*Практическая значимость: Почувстует ли рядовой клиент улучшения? Стоит ли это того, что 5% клиентов стали ждать значительно дольше?*

*Бизнес-эффект: Сколько стоила внедрение алгоритма и какую уже финансовую выгоду он принес?*